PREAPPS-502: View event details

pl12133 opened the pull request:
> As a user, I want to see the details of a event in my calendar when I hover over (desktop) or click on (mobile) it in any calendar view.

Depends on https://github.com/Zimbra/zm-api-js-client/pull/8
Depends on https://github.com/Zimbra/zm-x-ui/pull/5

There was a good amount of stuff to do for getting the `ModalDrawer` component to look consistent on mobile, such as moving the main application border off of the `<main>` tag and onto the respective navigation components.

TODO:
 - [x] c445b07f15b547eee716810393722a2d24e4a8d0 **critical** fix bug where adding a new event causes 100% CPU usage (infinite loop?).

pl12133 commented on src/components/mouse-tooltip/index.js:
```diff
@@ -0,0 +1,21 @@
+		},
+		transformOrigin: {
+			x: 0,
+			y: 144
```
The `main` element of the App is transformed by `144px`, so we must transform `-144px` on any fixed element within the `main` element to make it line up with the cursor correctly.

I spent some time trying to get the whole App to use flexbox instead of absolute positioning to fix this, but it seems like it's more trouble than it's worth.

dlesynacor replied:
@billneff79 @pl12133 seems kind of hacky, i can take a look at converting the App into cssgrid or flexbox and if anything we can just undo this after?

pl12133 replied:
@dlesynacor I would be all for that, I committed my attempt on the [`research/full-flexbox`](https://github.com/Zimbra/zm-x-web/tree/research/full-flexbox) branch, but ran into trouble with the scrolling regions in the Contacts vertical.

billneff79 commented on src/components/calendar/index.js:
```diff
@@ -426,14 +433,19 @@ export default class Calendar extends Component {
-		// }
+
+		const { matchesScreenMd } = this.props;
+		if (!matchesScreenMd) {
```
- [ ] click is for all environments, not just for mobile

pl12133 replied:
Done, this is handled in `src/components/calendar/event.js`.

billneff79 commented on scripts/apollo-codegen.js:
```diff
@@ -7,7 +7,7 @@
 
 const shell = require('shelljs');
 
-const ROOT = '../zm-x-api-js-client/src/schema';
```
We aren't using typescript at all in zm-x-web right now, so we can probably just delete any of the typescript related scripts/code/deps

billneff79 replied:
that's up to you though

pl12133 replied:
I'm still unsure about where the `apollo-codegen` piece fits into `zm-api-js-client`, but it seems like we have to leave it here if we want to be able to re-generate the `generated-schema-types`.

billneff79 replied:
The _only_ thing we need `apollo-codegen` for is to generate javascript consumable constant declarations from typescript enums.  e.g. when the graphql schema says that we have an enum of `Foo` with of `A`, `B`, or `C`, `apollo-codegen` creates a javascript file with a `Foo` exports that is an object with keys `A`, `B`, and `C` (probably with values the same as the key name).

Given that, it becomes clear that `apollo-codegen` is only useful if you have `typescript` files in your project and you want non-typescript consuming applications to be able to use your enum definitions.  Given that `zm-x-web` does not have any exports period, I think we can rip `apollo-codegen` and any related scripts out of the `zm-x-web` repo

billneff79 commented on scripts/apollo-codegen.js:
```diff
@@ -7,7 +7,7 @@
 const shell = require('shelljs');
 
-const ROOT = '../zm-x-api-js-client/src/schema';
+const ROOT = '../zm-api-js-client/src/schema';
```
the repo directory is more correct, but this is still not using it as an npm dep, which it would need to be if you were to actually use it

billneff79 commented on src/components/app-navigation/style.less:
```diff
@@ -6,6 +6,7 @@
 	width: 100%;
 	background-color: white;
 	overflow: visible;
+	border-bottom: solid 1px @gray-lighter;
```
Should use `@border-color-primary` (which is `@gray-lighter`), but is more semantically correct

billneff79 replied:
Did this not get  committed @pl12133 ?  The task is marked as complete but I'm still seeing `@gray-lighter` in the code

pl12133 replied:
Woops, 954d9400b529c7d873a64744ce6056d4d430e979 only addressed the mobile view. Will do this one as well:

 - [ ] `@border-color-primary` for `app-navigation`

billneff79 commented on src/components/calendar/event-details/event-details-tooltip.js:
```diff
@@ -0,0 +1,32 @@
+		return (
+			<MouseTooltip origin={origin} class={cx(style.eventTooltip, props.class)}>
+				<Localizer>
+					<button
```
We use this in so many places, I wonder if the close x should be a component or type of `Button`

pl12133 replied:
I tidied up the `<CloseButton>` component and used it here.

billneff79 commented on src/components/calendar/event-details/event-details.js:
```diff
@@ -0,0 +1,85 @@
+				<h2>{event.name}</h2>
+
+				{event.location && (
+					<div class={style.location}>at {event.location}</div>
```
Should `i18n` this

billneff79 commented on src/components/calendar/event-details/event-details.js:
```diff
@@ -0,0 +1,85 @@
+
+				<div>
+					<time>{`${format(new Date(event.start), 'ddd, MMM DD, hh:mm A')} - ${format(new Date(event.end), 'hh:mm A')}`}</time>
+					{<Icon size="xs" name={`fa:bell${event.alarm ? '' : '-slash'}`} />}
```
To match the mocks, you need some padding between the time and the reminder icon

billneff79 commented on src/components/calendar/event-details/event-details.js:
```diff
@@ -0,0 +1,85 @@
+
+				{!event.isOrganizer && event.organizer && event.organizer.address && (
+					<div class={style.organizer}>
+						<Text id="calendar.eventFields.organizer" />
```
Is this how organizer is supposed to display?  This mock shows it like:
`foo@bar.com (organizer)`
https://app.zeplin.io/project/59303d6aeda20fbc55c6c621/screen/5a56948c458063c2c7a3d6b4

pl12133 replied:
Just checked with @tstegner and he said it's fine to go with the `Organizer:` label as it is now.

billneff79 commented on src/components/calendar/event-details/style.less:
```diff
@@ -0,0 +1,83 @@
+@import '~@zimbra/x-ui/refs.less';
```
typically do `@import (reference)  '~@zimbra/x-ui/refs.less'`

pl12133 replied:
I believe the `refs.less` file will take care of this for us since it imports a `reference` to `variables.less`, in most places in the app there is a regular import instead of a `reference`.

billneff79 replied:
true...

billneff79 commented on src/components/calendar/event-details/style.less:
```diff
@@ -0,0 +1,83 @@
+	}
+
+	> hr {
+		margin: 12px 0 12px 0;
```
could shorthand to `margin: 12px 0`.  Might save  half a byte ;)

billneff79 commented on src/components/calendar/event.js:
```diff
@@ -73,20 +78,68 @@ export function CalendarEventWrapper(props) {
+	hideHoverTooltip = () => this.setState({ hoverOrigin: false })
+
+	componentWillMount() {
+		document.addEventListener('mousemove', this.handleMouseMove);
```
It seems unnecessary to track `handleMouseMove` for every calendar event on the page.  We only care about it if the hover tooltip is active, right?  Why not add the event listener when you show the event details and remove the event listener when you hide the event details?

pl12133 replied:
Changed over to `mouseleave`, I originally used `mousemove` because I expected to need a `Portal` here.

billneff79 commented on src/components/calendar/style.less:
```diff
@@ -119,7 +123,7 @@
 
 	.event {
-		z-index: 0;
+		// z-index: 0; // Why was this ever here?
```
delete instead of comment out

billneff79 commented on src/components/calendar/style.less:
```diff
@@ -128,19 +132,14 @@
-			font-weight: normal;
-			font-size: 100%;
-		}
+		// Does this select anything? Leaving as commented...
```
If we can find a thing it selects, we should remove it

billneff79 commented on src/components/mouse-tooltip/index.js:
```diff
@@ -0,0 +1,21 @@
+import { h, Component } from 'preact';
+
+export default class MouseTooltip extends Component {
```
By not using `popper` here, we are going to run into the same issues we had before it, with modals/tips rendering out of the viewport.  Is there a way we can use that here?

pl12133 replied:
I was having trouble adding `popper.js` here, so instead I changed the Tooltip to always open towards the center of a containing element, defaulting to `document.body`. This is similar to the way we position our context menus.

billneff79 replied:
I'm seeing the tooltip jump around a bit as it adds data to the contents.  Can you tell me what problems you were having with popper?  Maybe I can look into it.

billneff79 commented on src/components/skeletons/with-sidebar.js:
```diff
@@ -2,10 +2,21 @@ import { h } from 'preact';
+
+			{errors.length && (
+				<div class={s.main}>
+					<h2>The application has encountered an error</h2>
```
i18n

billneff79 commented on src/components/toolbar/style.less:
```diff
@@ -1,6 +1,10 @@
 	flex: 1;
 	background-color: white;
+	height: @mobile-toolbar-height;
+	border-bottom: 1px solid @gray-lighter;
```
`@border-color-primary` instead

billneff79 commented on src/components/recurrence/index.js:
```diff
@@ -0,0 +1,56 @@
+	return <span {...props}>{getFrequency(recurrence)}</span>;
+}
+
+function getFrequency(recurrence) {
```
You would make my day if you added some unit tests for this

pl12133 replied:
Will do:

 - [ ] unit tests for `getFrequency`

billneff79 commented on src/lib/util.js:
```diff
@@ -476,3 +476,8 @@ export function formatBytes(bytes, decimals) {
 	return parseFloat((bytes / Math.pow(k, i)).toFixed(dm)) + ' ' + sizes[i];
 }
+
+export function capitalizeFirstLetter(string) {
```
I would like it to become standard practice that we unit test new/modified library functions

pl12133 replied:
- [x] unit tests for `capitalizeFirstLetter`

pl12133 commented on src/components/calendar/index.js:
```diff
@@ -471,8 +483,13 @@ export default class Calendar extends Component {
+		this.clearNewEvent();
+	}
+
+	constructor(props) {
```
I think that would only be relevant if I needed to access `this.context` in the constructor, I don't usually pass context along to `super` and it hasn't caused a problem yet.

billneff79 replied:
Since we know the parent is not expecting/using `context` for anything that's correct.  We actually know that `super` is doing anything with `props` as well, we probably just have it because of the lint rule.  The lint rule should really catch a missing context as well, since props and context are the two args for the constructor.  You don't need to change anything, I more wanted to highlight it for our new folks.  These type of gotchas are why we tend to avoid `constructor` unless we really need to use it.

pl12133 commented:
Comments from JIRA:

 - [x] de433c968cbdb31e230b7fb70d97bbea5bca96e3 Meetings that span more than one day should show the ending date as well as the beginning date
 - [x] 8a4829c Long locations overflow the modal

billneff79 commented on tests/unit/lib/util.js:
```diff
@@ -40,4 +40,14 @@ describe('util', () => {
+
+	describe('capitalizeFirstLetter', () => {
+		it('should capitalize the first letter of a sentence', () => {
+			[ 'one', 'two', 'Three', 'Four', 'five' ].map(capitalizeFirstLetter).forEach((word, index, arr) => {
```
You don't need to change it, but I prefer not writing code that itself should be tested :)  Could have just written this as:
```
expect([ 'one', 'two', 'Three', 'Four', 'five' ].map(capitalizeFirstLetter)).to.eql(['One', 'Two', "Three', 'Four', 'Five']);
```

billneff79 replied:
I would probably add a two-token word as well, e.g. verify that `first word` gets changed to `First word`

pl12133 commented:
@billneff79 OK ripped out the `apollo-codegen` stuff, updated the `capitalizeFirstLetter` unit tests, and changed `@gray-lighter -> @border-color-primary`; should be all set.
